core/src/main/java/org/apache/druid/java/util/common/guava/ParallelMergeCombiningSequence.java

```diff
@@ -0,0 +1,1077 @@
+   * a set of {@link BatchedResultsCursor} are placed in a {@link PriorityQueue} to facilitate ordering to merge results
+   * from these cursors, and combine results with the same ordering using the combining function.
+   */
+  abstract static class BatchedResultsCursor<E>
+      implements ForkJoinPool.ManagedBlocker, Comparable<BatchedResultsCursor<E>>
+  {
+    final Ordering<E> ordering;
+    volatile ResultBatch<E> resultBatch;
+
+    BatchedResultsCursor(Ordering<E> ordering)
+    {
+      this.ordering = ordering;
+    }
+
+    public abstract void initialize();
+
+    public abstract void advance();
+
+    public abstract boolean isDone();
+
+    void nextBatch()
+    {
+      try {
+        ForkJoinPool.managedBlock(this);
+      }
+      catch (InterruptedException e) {
+        throw new RuntimeException("Failed to load next batch of results", e);
+      }
+    }
+
+    public void close()
+    {
+      // nothing to close for blocking queue, but yielders will need to clean up or they will leak resources
+    }
+
+    public E get()
+    {
+      return resultBatch.get();
+    }
+
+    @Override
+    public int compareTo(BatchedResultsCursor<E> o)
+    {
+      return ordering.compare(get(), o.get());
+    }
+
+    @Override
+    public boolean equals(Object o)
+    {
+      if (!(o instanceof BatchedResultsCursor)) {
+        return false;
+      }
+      return compareTo((BatchedResultsCursor) o) == 0;
+    }
+
+    @Override
+    public int hashCode()
+    {
+      return Objects.hash(ordering);
+    }
+  }
+
+
+  /**
+   * {@link BatchedResultsCursor} that wraps a {@link Yielder} of {@link ResultBatch} to provide individual rows
+   * of the result batch.
+   */
+  static class YielderBatchedResultsCursor<E> extends BatchedResultsCursor<E>
+  {
+    final SequenceBatcher<E> sequenceYielder;
+    Yielder<ResultBatch<E>> yielder;
+
+    YielderBatchedResultsCursor(SequenceBatcher<E> sequenceYielder, Ordering<E> ordering)
+    {
+      super(ordering);
+      this.sequenceYielder = sequenceYielder;
+    }
+
+    @Override
+    public void initialize()
+    {
+      yielder = sequenceYielder.getBatchYielder();
+      resultBatch = yielder.get();
+    }
+
+    @Override
+    public void advance()
+    {
+      if (!resultBatch.isDrained()) {
+        resultBatch.next();
+      }
+      if (resultBatch.isDrained() && !yielder.isDone()) {
+        nextBatch();
+      }
+    }
+
+    @Override
+    public boolean isDone()
+    {
+      // yielder will never produce a 'terminal' result batch, so only check that we drain the final batch when the
+      // yielder is done
+      return resultBatch == null || (yielder.isDone() && resultBatch.isDrained());
+    }
+
+    @Override
+    public boolean block()
+    {
+      if (yielder.isDone()) {
+        return true;
+      }
+      if (resultBatch == null || resultBatch.isDrained()) {
+        final Yielder<ResultBatch<E>> nextYielder = yielder.next(resultBatch);
```

Comment:
It seemed more efficient to re-use the same batch since it would result less garbage objects and less pressure on the garbage collector, and not particularly less readable to me at least. Can I flip the question, is there a good reason to make a new object when I have a perfectly good one already allocated and ready to go?

Reply:
After thinking more about this, I agree it is a bit more obvious to use a new batch. I'll run benchmarks to make sure there isn't any performance degradation and make the change.

Reply:
My main concern was about if it's violating any concurrent access. I think there's no concurrency issue but creating new batches will be easier to find/remove concurrency issues if they are added in the future. But if there was any performance issue, then it would make more sense to reuse the same instance.

Reply:
I'm also wondering if the performance can be better if `ResultBatch` is reused and annotated with `@Contended`. Apparently, contended objects are allocated more space and so they can be isolated without polluting cache lines.

This will need some experiments, so I don't think my comment should be blocker of this PR.

Reply:
There appears to be no performance issues with creating a new object, so I will change it to be more in spirit of how the yielder is working, where a new one is returned by next.